### DISCUSSION TITLE
导出有道词典单词本

### DISCUSSION THREAD

**Discussion by LeetaoGoooo** [LeetaoGoooo/discussion-blog#223] [随笔]
## 更新 2023-10-26

经由 @XYZ-99 提示才发现网页默认单词列表的只是【**我的单词本**】（猜测应该是默认的单词本，手机端可以设置），对于多单词本的情况下脚本无法一次性导出，所以对网页有道词典再次研究，发现了获取单词本的接口，以及对现有的接口的推测，对原有导出脚本再次升级了，现导出脚本支持**导出所有的单词本的所有单词**，项目地址 [export-youdao](https://github.com/LeetaoGoooo/export-youdao)

# 前言

有道词典功能太多了，奈何我只想要一个查单词、背单词然后没有广告的软件，所以最后决定投入欧陆词典的怀抱，迁移到过程中突然发现竟然 Macos 和 IOS 的有道词典没有导出单词本的功能，只好自己动手了。

# 导出脚本

```python
import browser_cookie3
import requests
import time
import random
from typing import TypedDict, List

class Book(TypedDict):
    bookName: str
    bookId: str


cj = browser_cookie3.chrome()


def get_books() -> List[Book]:
    """
    获取所有的单词本列表
    """
    url = "https://dict.youdao.com/wordbook/webapi/v2/opts"
    resp = requests.get(url=url, cookies=cj)
    if resp.status_code != 200:
        print("请求有道单词本失败,请确保已经登录有道单词本: https://www.youdao.com/webwordbook/wordlist")
        return []
    response = resp.json()
    print("获取单词本列表成功,累计获取{}个单词本".format(len(response["data"]["book"])))
    return response["data"]["book"]


def get_words_by_book(limit, offset,sort, book:Book):
    url = "https://dict.youdao.com/wordbook/webapi/v2/word/list"
    resp = requests.get(url=url,params={
        "limit": limit,
        "offset": offset,
        "sort": sort,
        "lanTo": None,
        "lanFrom": None,
        "bookId": book["bookId"],
    }, cookies=cj)
    if resp.status_code != 200:
        print("请求有道单词本失败,请确保已经登录有道单词本: https://www.youdao.com/webwordbook/wordlist")
        return [], 0
    
    response = resp.json()
    total = response["data"]["total"]
    words = response["data"]["itemList"]
    return [word["word"] for word in words], total


def export_book_words(book:Book, limit=48, offset=0, sort="time"):
    words, total = get_words_by_book(limit, offset, sort, book)

    with open("words.txt", "a+") as f:
        f.write("\n".join(words))
        f.write("\n")
    
    offset += limit
    
    print("当前导出单词本【{}】 进度: {}/{}".format(book["bookName"], total if offset > total else offset, total))

    if offset  > total:
        print(f"导出单词本【{book['bookName']}】完成")
        return
    
    sleep_time = random.randint(1, 3)
    print("休息{}秒".format(sleep_time))
    time.sleep(sleep_time)

    return export_book_words(book, limit, offset, sort)    

def export_words_to_txt():
    books = get_books()
    if not books:
        return
    for book in books:
        print("单词本: {}".format(book["bookName"]))
        export_book_words(book)
        
    

if __name__ == "__main__":
    export_words_to_txt()
```
 
# 运行截图

<img width="337" alt="image" src="https://github.com/LeetaoGoooo/discussion-blog/assets/10162120/e05cc8a4-e4c9-47e9-abcb-cdbad0207aa5">

最终导出一个 txt 文件，然后在欧路词典的网页端 https://my.eudic.net/studylist/import/  进行导入

<img width="1000" alt="image" src="https://github.com/LeetaoGoooo/discussion-blog/assets/10162120/419957e8-d136-4b3f-96f6-659f9660632f">

**Comment by Dieinwarm** [comment]
我觉得背单词 百词斩 也挺好用的啊

**Reply by LeetaoGoooo** [reply]
背单词百词斩是蛮好的，我这个主要是查单词的～😄

**Comment by XYZ-99** [comment]
是不是这个API只能获取默认单词本里的，我找了半天没找到查询其他手动添加的组别里单词的办法，看来只能在电脑端批量操作一下再爬了嘛

**Comment by LeetaoGoooo** [comment]
> 是不是这个API只能获取默认单词本里的，我找了半天没找到查询其他手动添加的组别里单词的办法，看来只能在电脑端批量操作一下再爬了嘛

应该可以获取其它单词本的，我今天把如何获取其它单词本的方法在文章补全一下

**Reply by XYZ-99** [reply]
震惊于站主的回复之快以及起床之早，并对站主的热心表示感谢，且对站主这份虽无人问津但与时俱进排版简明的帖子表达赞美与支持（抱歉进行了一些话痨输出，总之就是感谢！）

**Reply by LeetaoGoooo** [reply]
夸张了，夸张了，你这比我起的更早啊～

**Reply by LeetaoGoooo** [reply]
已经更新脚本了～～

**Reply by XYZ-99** [reply]
太高效了！哈哈哈误会了，我不是北京时间；有一个问题不知可不可以请教一下，我之前本来打算抓包看看这 API 长什么样，但是网页版甚至都不能调到非默认单词本，于是就放弃了。所以想请教一下，站主是对移动端抓包了之后得到的这些 API 吗？还是有什么其他更方便的方法找到的这些 API

**Reply by LeetaoGoooo** [reply]
就是网页版的 Api ，其实网页端其实已经把接口都暴露出来了，只是在网页上没有使用而已，然后结合现有的接口大胆推测了一下，幸运是，接口按照预期生效了

**Comment by burial12** [comment]
你好！求问怎么弄啊，我为了来提问特地来注册了一个账号，我不懂程序，你给的link点过去以后，怎么copy这个有道单词本的cookie啊，网页版登陆进去没有单词本，我是mac电脑，用的是有道桌面版apple store下载的，求指点，有道怎么这么缺德，苹果电脑手机都不能导出单词本，气死我了啊。

**Reply by LeetaoGoooo** [reply]
你先在这个页面登陆 https://www.youdao.com/ ，然后再打开 https://www.youdao.com/webwordbook/wordlist 这个页面就能看到单词本了，cookie 的话，你之前使用文章里的脚本，用 pip 先安装 browser-cookie3 和 requests 库，直接运行脚本，就不要管 cookie 的问题了，程序会让你输入密码，然后自动获取浏览器的 cookie；

担心这个不安全的话，就使用 [export-youdao](https://github.com/LeetaoGoooo/export-youdao) 这个项目，然后需要打开控制台 fn + 12 点到 network 的 tab 页，然后

![image](https://github.com/LeetaoGoooo/discussion-blog/assets/10162120/423f7063-744b-4d80-a876-21ca5ced33dc)

把 cookie 复制，按照项目说明，就可以导出了，有其他任何问题的话，可以随时留言

**Reply by burial12** [reply]
哇你真的好好，我就怕我搞不定，我在英国，等我先试试看不行再来求教，先感恩致谢！！！！

**Reply by burial12** [reply]
运行脚本在哪里运行啊，我成功复制cookie了，然后打开你的那个export程序点进去看见那个脚本复制+我的cookie以后在哪里去运行呀

**Reply by LeetaoGoooo** [reply]
> 运行脚本在哪里运行啊，我成功复制cookie了，然后打开你的那个export程序点进去看见那个脚本复制+我的cookie以后在哪里去运行呀

需要一点 Python 基础，你把脚本复制下来，新建一个 export.py 的文件，把代码复制进去，安装 对应的依赖，然后就可以在命令行下运行 python export.py

可以简单看一下这个视频 【Python 导出网易有道词典单词本-哔哩哔哩】 https://b23.tv/QR9ycbH

**Reply by LeetaoGoooo** [reply]
有成功导出吗？

**Reply by burial12** [reply]
我选择了一个更简单的方法，让window系统的人登陆我的账号下载了词典，15分钟搞定。你的这个方法对毫无计算机背景的我来说花的时间太多了，没空啊，学习已经很累了，连我对门计算机专业的人都懒得帮我配环境（他说的这个词，我都不知道啥意思，哈哈），不过还是多谢，以后等他有空，我多少得让他帮我弄好了在我的mac电脑上运行一下

**Reply by LeetaoGoooo** [reply]
haha ~ 这样子的话我还是写个应用程序给后面想用的人用吧

**Reply by burial12** [reply]
感恩！！麻烦了，拜托了！也不知道有道为什么这么坑，你有空写一个吧，我第一个下载

**Reply by LeetaoGoooo** [reply]
已经写了一个 GUI 的版本，你可以试试～ https://github.com/LeetaoGoooo/WordsOut/releases/tag/v0.0.1

**Reply by burial12** [reply]
文件已经损坏

**Reply by LeetaoGoooo** [reply]
可以了～～不过黑暗模式的时候界面有点问题～～

**Reply by burial12** [reply]
还是不行哎，我没用黑暗模式，就是普通mac，没关系啦，我已经导出2000个单词了暂时不用这个
